### PR TITLE
[build] Build Catapult with Visual Studio (MSVC) and Vcpkg

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
## What is the current behavior?
* Actual build process is a bit uncomfortable for windows users on VS

## How have you changed the behavior?
* This PR provides an additional build mode for Catapult which leverages a seamless integration with VS build process using Vcpkg as package manager.

## How was this change tested?
* There are no changes in managed code and all modified files have an impact only on builds.
* Previously existing build methods remain untouched and keep working as expected
* Test methodology: binaries are correctly generated and work on Windows

## Requirements
* No additional component requirements introduced. Vcpkg affects only the build toolchain

### CODE REVIEW

File: client/catapult/src/catapult/cache/CacheChangesSerializer.h [hunk 1]
```diff
@@ -25,6 +25,11 @@
 #include "catapult/io/Stream.h"
 #include <vector>
 
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4702) /* unreachable code */
+#endif
```
Comment (Wayonb) [inline client/catapult/src/catapult/cache/CacheChangesSerializer.h:31]: @Jaguar0625 Do you know why this warning is not seen with Conan build? Will need to look into this. 🤔

**Review by Wayonb — CHANGES_REQUESTED** [review]
Nice Job with a few comments. 👍🏾 

Also, for MongoDB (c/c++) drivers, we have certain features turned off.
I noticed that ZLIB was built for the Mongo-c-driver. We usually disable ZLIB, but it does not seem to be allowed in the vcpkg package.
https://www.mongodb.com/docs/languages/c/c-driver/current/libmongoc/ref/packages/#vcpkg-optional-features

I will need to rebuild catapult deps again to see what is enabled.

File: client/catapult/CMakeGlobalSettings.cmake [hunk 2]
```diff
@@ -123,6 +123,9 @@ if(MSVC)
 	set(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} /DEBUG:FASTLINK")
 	set(CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} /DEBUG")
 
+    add_compile_options(/MP)            # Enable parallel compilation
+    add_compile_options(/GA)            # Optimizes for Windows applications
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/CMakeGlobalSettings.cmake:127]: Sure

File: client/catapult/CMakeSettings.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      "generator": "Visual Studio 17 2022 Win64",
+      "configurationType": "Release",
+      "inheritEnvironments": [ "msvc_x64_x64" ],
+      "buildRoot": "${projectDir}\\build\\${name}",
+      "installRoot": "${projectDir}\\build\\install\\${name}",
+      "ctestCommandArgs": "",
+      "cmakeToolchain": "${env.VCPKG_ROOT}\\scripts\\buildsystems\\vcpkg.cmake"
+    }
+  ]
+}
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/CMakeSettings.json:34]: This file is generally updated by VS ui which, on save, does not append the final new line.
To amend this is necessary to remember to reopen the file in "Edit Json" mode and adjust the file.
I will try to remember for the future.

image

File: client/catapult/docs/BUILD-conan.md [hunk 4]
```diff
@@ -62,24 +81,26 @@ Following instructions should work on Mac, Linux (Ubuntu 20.04) and Windows.
 While Conan will be building and installing packages, you might want to go for a ☕ (or lunch),
 as this will probably take *a bit*.
 
+### Install conan sources repo and get catapult source code (any OS)
 ```sh
-conan remote add nemtech https://conan.symbol.dev/artifactory/api/conan/catapult
-
-git clone https://github.com/symbol/symbol.git
-cd symbol/client/catapult
+$> conan remote add nemtech https://conan.symbol.dev/artifactory/api/conan/catapult
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-conan.md:86]: As you pointed out is only a stylistic reason.
As a side note I generally do not encourage copy paste of the command line as it may happen some arguments could be derived from specific local environment variables or paths.

In any case I'm removing those to be consistent.

File: client/catapult/docs/BUILD-conan.md [hunk 5]
```diff
@@ -27,12 +31,27 @@ Following instructions should work on Mac, Linux (Ubuntu 20.04) and Windows.
 
      Run all commands from a command prompt that has access to Visual Studio and Git. This can be accomplished by using the "Native Tools Command Prompt" shortcut installed by Visual Studio on the Start Menu.
 
-  2. Install latest version of [Conan](https://conan.io/downloads.html).
+  2. Install latest version of [Conan](https://conan.io/downloads.html). You can proceed in two ways:
+  
+        - if you have Python installed, you can use pip:
+		
+		```shell
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-conan.md:38]: Sure

File: client/catapult/docs/BUILD-conan.md [hunk 6]
```diff
@@ -27,12 +31,27 @@ Following instructions should work on Mac, Linux (Ubuntu 20.04) and Windows.
...
+        $> pip install conan
+        ```
+        _Even on recent versions of python it has been reported that this command installs conan version 1.x: for this reason, before proceeding further, it is necessary to check version of conan installed by `conan -v`. Should the reported version be lower than 2.x please run `pip install conan --upgrade`. Eventually verify the new installed version is 2.x_
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-conan.md:41]: Yes sure.

File: client/catapult/docs/BUILD-vs.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+[Git]: https://git-scm.com/download/win
+
+# Building with VisualStudio 2019
+
+Following instructions are meant for integrated builds using VisualStudio on Windows only. This instructions **do not appy to VSCode**
+
+## Mandatory version requirements
+- [Git] version >= 2.25
+- [Python] version >= 3.6
+- [Visual Studio] 2017+ (2019 recommended): tested on 2022 (17.9.6)
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-vs.md:13]: Ok.

File: client/catapult/docs/BUILD-vs.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+- [vcpkg] package manager >= 2024.04.26 (instructions for install below)
+
+**Important: Ensure your VS setup includes CMAKE support**
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-vs.md:16]: Actually no. That tutorial explains how to install vcpkg and integrate it in VS for any project.
In this PR we use vcpkg integrated in manifest mode and beside cloning and boostrapping vcpkg no additional steps are required.
I believe explaining the full mechanics of vcpkg might be beyond the scope of this PR and of the project in general

File: client/catapult/docs/BUILD-vs.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+### Obtaining the source code
+From a terminal window (or using Git for Windows if you have it) clone the Symbol repository
+```shell
+git clone https://github.com/symbol/symbol.git
+```
+
+## Build Catapult
+
+1. Open your [Visual Studio] instance.
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-vs.md:44]: Yes it is intended. At the top of the file is defined a set of default links which are set by default every time a label is enclosed in square brackets (but without the following link in braces). So every time `[Visual Studio]` is encountered the macro picks `[Visual Studio]: https://www.visualstudio.com/downloads` and replaces the label to `[Visual Studio]( https://www.visualstudio.com/downloads)`

File: client/catapult/docs/BUILD-conan.md [hunk 10]
```diff
@@ -62,24 +81,26 @@ Following instructions should work on Mac, Linux (Ubuntu 20.04) and Windows.
...
-cd build/Release
+$> conan install . --build=missing -s build_type=Release
+$> cd build/Release
 ```
 
 ### Windows
-```shell
-conan install . --build=missing -s compiler.cppstd=17 -s build_type=Release
-cd build
+```sh
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/BUILD-conan.md:98]: I believe the rendering is mostly the same.
Here the list of main _console_ lexers and their aliases
```
Shell:      console, shell
Bash:       bash, sh, zsh
PowerShell: powershell, ps
DOS:        dos, bat, cmd
```

Will be consistent with `sh` though: after all the build with conan is for any OS

File: client/catapult/docs/README.md [hunk 11]
```diff
@@ -53,13 +53,15 @@ Windows:
 
 * Visual Studio 2017 (15.8)
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply client/catapult/docs/README.md:54]: Removing

**Comment by AndreaLanfranchi** [comment]
> Nice Job with a few comments. 👍🏾
> 
> Also, for MongoDB (c/c++) drivers, we have certain features turned off. I noticed that ZLIB was built for the Mongo-c-driver. We usually disable ZLIB, but it does not seem to be allowed in the vcpkg package. https://www.mongodb.com/docs/languages/c/c-driver/current/libmongoc/ref/packages/#vcpkg-optional-features
> 
> I will need to rebuild catapult deps again to see what is enabled.

It seems like the [vcpkg portfile definition](https://github.com/microsoft/vcpkg/blob/bd2b54836beed96e1efbe9aaf8ee800f5448856d/ports/mongo-c-driver/portfile.cmake#L55C23-L55C29) provides an option for it bound to a `SYSTEM` tag which I'm not sure which meaning has. Looking around.

File: client/catapult/src/catapult/cache/CacheChangesSerializer.h (hunk 1 above)
Comment (Jaguar0625) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:31]: i don't ... i wonder what file(s) are including this when warning is raised 🤔

Comment (AndreaLanfranchi) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:31]: It might be possible is something surfaced by the application of the parallel build flag (/MP) which is now enabled for MSVC. I will double check

Comment (AndreaLanfranchi) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:31]: Actually also the Conan build (Windows + MSVC) triggers the error. Here the extract of the log

```
        C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\include\stdfloat(12): warning STL4038: The contents of <stdfloat> are available only with C++23 or later. [D:\GitHub\symbol\client\catapult\build\tests\ca
       tapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\include\stdfloat(12): warning STL4038: The contents of <stdfloat> are available only with C++23 or later. [D:\GitHub\symbol\client\catapult\build\tests\ca
       tapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\include\stdfloat(12): warning STL4038: The contents of <stdfloat> are available only with C++23 or later. [D:\GitHub\symbol\client\catapult\build\tests\ca
       tapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         D:\GitHub\symbol\client\catapult\src\catapult\cache\ChunkedDataLoader.h(54,1): warning C4702: unreachable code [D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         D:\GitHub\symbol\client\catapult\src\catapult\cache\CacheChangesSerializer.h(39,1): warning C4702: unreachable code [D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         D:\GitHub\symbol\client\catapult\src\catapult\cache\CacheChangesSerializer.h(60,1): warning C4702: unreachable code [D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery.vcxproj]
         D:\GitHub\symbol\client\catapult\src\catapult\cache\CacheChangesSerializer.h(61,1): warning C4702: unreachable code [D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery.vcxproj]
```

Which causes the build process to terminate with 
```
       "D:\GitHub\symbol\client\catapult\build\ALL_BUILD.vcxproj" (destinazione predefinita) (1) ->
       "D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery.vcxproj" (destinazione predefinita) (208) ->
       (destinazione: ClCompile) ->
         D:\GitHub\symbol\client\catapult\src\catapult\cache\ChunkedDataLoader.h(54,1): error C2220: the following warning is treated as an error [D:\GitHub\symbol\client\catapult\build\tests\catapult\local\recovery\tests.catapult.local.recovery
       .vcxproj]

    Warnings: 178
    Errors 1
```

**Comment by AndreaLanfranchi** [comment]
Just added commit 61e28e1d amending an error on MSVC buillds.

This
```
add_definitions(-D_WIN32_WINNT=0x0601 /w44287 /w44388)
```
is wrong as it mixes CMake definition with compiler warning settings.

Changed to 
```
	add_definitions(-D_WIN32_WINNT=0x0601)

	add_compile_options(/w44287)		# 'operator' : unsigned/negative constant mismatch
	add_compile_options(/w44388)		# 'token' : signed/unsigned mismatch
```

This was also the reason Ninja generator produced an error.

File: client/catapult/src/catapult/cache/CacheChangesSerializer.h (hunk 1 above)
Comment (Wayonb) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:31]: Maybe the issue has to do with MSVC.  I removed the disable of the unreachable code warning and built with  ``msbuild /p:Configuration=Release /p:Platform=x64 /m ALL_BUILD.vcxproj`` as documented and did not have any errors.  It might be interesting to see what the MSVC command is🤔

Comment (AndreaLanfranchi) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:33]: Well `msbuild` actually **is** msvc: is a CLI shortcut for the same build command VS invokes from the UI.

TL;DR: The issue is related to MSVC builds in **Debug** mode.

It's been my fault not to specify I've been issuing `Debug` builds (not Release) also for conan.  @Wayonb test has been carried out for Release. And I can confirm this PR in Release mode does not trigger the unreachable code either. In Debug mode, instead, additional static analysis is performed over the executed code. This leads to the issue being triggered when tests are built in those cases when classes are initialized with literal constants: the analisys of the compiled test surfaces some paths are not reachable.

Example from ChunckedDataLoader.h:
```
		/// Loads the next data chunk of at most \a numRequestedEntries into \a destination.
		void next(uint64_t numRequestedEntries, typename TStorageTraits::DestinationType& destination) {
			numRequestedEntries = std::min(numRequestedEntries, m_numRemainingEntries);
			m_numRemainingEntries -= numRequestedEntries;
			while (numRequestedEntries--)
				TStorageTraits::LoadInto(TStorageTraits::Load(m_input), destination); // <-- Not reachable if m_numRemainingEntries is known at compile time to be zero.
		}
```

I've just changed the guard to constrain the warning suppression on debug builds only.

Comment (Wayonb) [reply client/catapult/src/catapult/cache/CacheChangesSerializer.h:33]: @AndreaLanfranchi Thanks for the information.  I was also able to repo the build failure in Debug.  I opened an issue to look into this https://github.com/symbol/symbol/issues/1051

File: client/catapult/vcpkg.json [hunk 12]
```diff
@@ -0,0 +1,678 @@
...
+    {
+      "name": "boost-container",
+      "version": "1.83.0#0"
+    },
+    {
+      "name": "boost-container-hash",
+      "version": "1.83.0#0"
+    },
+    {
+      "name": "boost-context",
```
Comment (Wayonb) [inline client/catapult/vcpkg.json:130]: Catapult does not need all the boost libs like ``context, contract, etc``.
Here is a list of what is removed - https://github.com/symbol/symbol/blob/b3c472b6b36eaf31e5f1d939ceb5b9b7e0f1186d/jenkins/catapult/baseImageDockerfileGenerator.py#L11

**Review by Wayonb — CHANGES_REQUESTED** [review]
I missed this file on the first go 😅

File: client/catapult/vcpkg.json (hunk 12 above)
Comment (AndreaLanfranchi) [reply client/catapult/vcpkg.json:130]: A bit of explanation is needed here.
The list of Catapult dependencies is in the section "dependencies" of that json file.
What you find in the section "overrides" has multipurpose:

1. The way vcpkg segments the Boost library is a bit different and more granular than "normal" boost libs
2. Many of the boost components require ancillary components which are not explicitly indicated in catapult dependencies
3. To ensure all components  of the boost set are built on the same version of the library I simply grabbed ALL possible boost targets built by vcpkg enforcing on them (in overrides section) the specific version 1.83

For instance this is the list of Boost's components built as dependencies of the Catapult's boost required components
```
1> [CMake]   * boost-algorithm:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-algorithm\064c9bd0deddb764f57008b07c0d42af035c1a50
1> [CMake]   * boost-align:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-align\6312ca6cedb77e48b123aa19e7ef98f5401b4020
1> [CMake]   * boost-any:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-any\8995a49c0eb70b1e306323f70d79dc5b6343bd6b
1> [CMake]   * boost-array:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-array\98d20d502bbd77d5a4312254bd47d83c55b4bbac
1> [CMake]   * boost-asio:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-asio\0d4a6f6c528811c3c79ac046e738e5bab3db995f
1> [CMake]   * boost-assert:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-assert\058c55df927f8a8180003c576bb4b536a46cf3e8
1> [CMake]   * boost-bind:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-bind\b846d10f5df5d2d6f27bd2fe13170cf3c09d8cbf
1> [CMake]   * boost-build:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-build\bd9d972c221f9e5bdd839246960791cdcfe80524
1> [CMake]   * boost-concept-check:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-concept-check\ec7a65ff7072399f9a926316cac0b1c6162d4ee5
1> [CMake]   * boost-config:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-config\26e9082aeb3829b11d9f0ea82de37db9a93b6620
1> [CMake]   * boost-container:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-container\06dc49b47f365371fd98d0373e46369bb97056c8
1> [CMake]   * boost-container-hash:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-container-hash\058bb680534a2375ec78dd366d9969967bd7570e
1> [CMake]   * boost-context:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-context\2c0ccde97a196bd6c78f0fbb52ae9c8b33a525b5
1> [CMake]   * boost-conversion:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-conversion\45d1d99389896f68454518ac9e7d516c88598fcd
1> [CMake]   * boost-core:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-core\736b9319578faba560f1bccd62ae53421c290ac4
1> [CMake]   * boost-coroutine:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-coroutine\d8a5c2cff3de497580fc8dc363ebc3e8ec0c9a1d
1> [CMake]   * boost-describe:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-describe\5287728efcb90f341cbdbf3e40edd2568baa3d98
1> [CMake]   * boost-detail:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-detail\9a596ae27e05912ce6ac9430ad0a2c60b0874517
1> [CMake]   * boost-dynamic-bitset:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-dynamic-bitset\783e8e0c0387114a445ad62dacbbb2f2ce910cd2
1> [CMake]   * boost-endian:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-endian\5321b25b38e55ebe1bbeb54242c104e208b643f8
1> [CMake]   * boost-exception:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-exception\8dce130433b2a5fc44ee576f9b39bb97d916661e
1> [CMake]   * boost-format:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-format\8b4697f0c1e3cc49f4bf08c12b748544c132db48
1> [CMake]   * boost-function:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-function\f066c30c0481a6f75452d3a1372ea3425adea156
1> [CMake]   * boost-function-types:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-function-types\77f27d230b3784277075862bba2c282977ea1f0a
1> [CMake]   * boost-functional:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-functional\aba27201e0794309bf68a44e8727378345927b38
1> [CMake]   * boost-fusion:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-fusion\0475f7220a08b37bfbea03b873211f9ea76fac08
1> [CMake]   * boost-integer:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-integer\f70c3d065ad54b0b64e16541cfdd86aceec513f3
1> [CMake]   * boost-interprocess:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-interprocess\685634ef2e0b7c514fc454d668e27c17937dc005
1> [CMake]   * boost-intrusive:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-intrusive\05886d36db2ae6ecb44cf3b964c2d0a4d486c129
1> [CMake]   * boost-io:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-io\f5d84659eada088f612945a42ae940ee7dcc37be
1> [CMake]   * boost-iterator:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-iterator\8870bee36b1dee543b554c3e0264e6871a6ad6e2
1> [CMake]   * boost-lexical-cast:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-lexical-cast\dde0974bedf5d7539835eeadb3183b72d0df75f0
1> [CMake]   * boost-modular-build-helper:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-modular-build-helper\843c4331dc1453c80f99e25be4513c99074c5cf2
1> [CMake]   * boost-move:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-move\3f4770bf9904c733be328e32d4017a0ed1157d2d
1> [CMake]   * boost-mp11:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-mp11\68c2556b43afbb853b5e724eb0fd5269a92088f1
1> [CMake]   * boost-mpl:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-mpl\dfcb80d8d1b767426e6714d6ba1b6a1473c0cc23
1> [CMake]   * boost-multi-index:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-multi-index\7d40ee58c3e54b9c841abc6dd57033366218f064
1> [CMake]   * boost-numeric-conversion:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-numeric-conversion\c3a65f3521df33b8a76605097a3046795f1107e7
1> [CMake]   * boost-optional:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-optional\145809a6e99a8ba579c3c0bef2c48969555a2bd5
1> [CMake]   * boost-parameter:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-parameter\1dce1428072cad1f8bc7b6c02f14cc6ba12da3fd
1> [CMake]   * boost-phoenix:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-phoenix\a4c5832e8ebf68cd4f2ee601a10aa7adfc12e6e9
1> [CMake]   * boost-pool:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-pool\e5c81fe62cb0da4fdc940d21d87837678a2e67cf
1> [CMake]   * boost-predef:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-predef\8e76add266584a1177e9a0f2c59c43a78d88e578
1> [CMake]   * boost-preprocessor:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-preprocessor\07609446e5cce5a0d9afed4570e0664708444b54
1> [CMake]   * boost-property-tree:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-property-tree\caa9ef5866c4dcf2869fca47fb3a478f74e07f7a
1> [CMake]   * boost-proto:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-proto\7bb3591b035347ce419b85ae321c0e1b6c71084a
1> [CMake]   * boost-random:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-random\045960cb6d4295d363e1748931ea56fc245f7c09
1> [CMake]   * boost-range:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-range\7d5d869d86b0abdd19022ee0b0c5f24192769291
1> [CMake]   * boost-ratio:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-ratio\0fafb812427bea7d49dd318f3e72778149f443ab
1> [CMake]   * boost-rational:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-rational\55ec23b1953b2e44dc43fda4547cf0b96bfe140f
1> [CMake]   * boost-serialization:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-serialization\7ebe53455e9721a76ebe14c651a2deafbcb758b8
1> [CMake]   * boost-smart-ptr:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-smart-ptr\81506ee1238bf93218cef84de86dca96dd0b3682
1> [CMake]   * boost-spirit:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-spirit\e2ebc2bcf72d8ae953ac54163a7c5b43f17396f0
1> [CMake]   * boost-static-assert:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-static-assert\50eac53fdc886f9de87fc06717a4ae3a939b28e9
1> [CMake]   * boost-throw-exception:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-throw-exception\22bf09f8d91024cc1e74b19bb6333baeea262b69
1> [CMake]   * boost-tokenizer:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-tokenizer\05ada2a1cc63fe9b4f5b4e9ac1d6ec63961e281b
1> [CMake]   * boost-tuple:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-tuple\a5a7d3cf0427278418ddc0e91a3858218535c8a9
1> [CMake]   * boost-type-index:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-type-index\58d41ff7ca046c8997ddd3f6c8ab801291edd9c4
1> [CMake]   * boost-type-traits:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-type-traits\6595fdcc50c1d1d5587ade48484a0b074ddb079b
1> [CMake]   * boost-typeof:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-typeof\ec85052de49f1335b474cee0689971b9ad4df92c
1> [CMake]   * boost-uninstall:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-uninstall\1dfbff4a218e77e371ebc397f0573859136c7c92
1> [CMake]   * boost-unordered:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-unordered\d78bed13cd960d7723ac5b0c82d2223c321ee0a9
1> [CMake]   * boost-utility:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-utility\175e91f86d24ff49330e7f5f95db441ceb85886e
1> [CMake]   * boost-variant:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-variant\60ca10585ea590b82692a0f715cfcc596d4799db
1> [CMake]   * boost-variant2:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-variant2\a249f8469b4c3ccb2aee3ece628d69a71d54b35f
1> [CMake]   * boost-vcpkg-helpers:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-vcpkg-helpers\18a50fedd4344db5ce1f82546be1889215381f78
1> [CMake]   * boost-winapi:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-winapi\cfd68f543033ed6b8cee9235ef287b7fd4f8709a
1> [CMake]   * boost-xpressive:x64-windows@1.83.0 -- E:\vcpkg\buildtrees\versioning_\versions\boost-xpressive\55de0dfb037123cd093c56acf61b22948e98295b
```

If it wasn't so picture this: I only specify we need boost's `chrono` at version 1.83 but it also requires (as internal dependency) say, `boost-numeric-conversion`. As version for a component is not propagated by vcpkg to other deps it would happen we get `chrono` at version 1.83 but linked to `numeric-conversion` at version 1.84 (which is boost's latest).

Comment (Wayonb) [reply client/catapult/vcpkg.json:130]: Thanks for the explanation. It is weird to link the dependency module from another version.

**Review by Wayonb — APPROVED** [review]
👍🏾

File: client/catapult/vcpkg.json (hunk 12 above)
Comment (AndreaLanfranchi) [reply client/catapult/vcpkg.json:130]: This better shows what I'm talking about
https://vcpkg.link/ports/boost-atomic/v/1.83.0/0

If not specified in the override Boost's assert would have been built on version 1.85 (latest)

**Comment by AndreaLanfranchi** [comment]
Re-requested review due to additions

File: client/catapult/CMakeLists.txt [hunk 13]
```diff
@@ -1,4 +1,19 @@
 cmake_minimum_required(VERSION 3.23)
+
+find_package(Git QUIET)
+if(NOT GIT_FOUND)
+    message(FATAL_ERROR "Git not found")
```
Comment (Wayonb) [inline client/catapult/CMakeLists.txt:5]: When would git not be found? When the source code is copied?

**Review by Wayonb — APPROVED** [review]
👍🏾
